### ISSUE TITLE
Dynamically add imports react-intl/locale-data/<locale>

### ISSUE THREAD

**Issue by SanthoshHari912** [yahoo/react-intl#777]
Hello, 
 Am using React-Intl for internationalisation in my React App. Here I have to support many locales which I don't know upfront. Am exposing a json folder where the user will add the locale specific json files.
For formatting numbers and data, am not sure how will I be able to import locales and add it to addLocaleData dynamically in React Class.

 I have to handle this in React Component Class itself. In my requirement, may not ask the user to include the locales explicitly in html files. Kindly suggest me any workarounds.

I would suggest to have an API similar to addLocaleData() which will accept the locale and import the locale implicitly. This enables the user to add the locale dynamically, without importing the locales in prior.

Regards
Santhosh

**Comment by ericf** [comment]
> I would suggest to have an API similar to addLocaleData() which will accept the locale and import the locale implicitly. This enables the user to add the locale dynamically, without importing the locales in prior.

Can you please explain this more? How would this function know where to load the locale data, and how would this work in the browser?

Is this a function that you could write that does what you need and calls `addLocaleData()`?

**Comment by SanthoshHari912** [comment]
Hi eric,

 Am creating a react based UI util library and for internationalization I use React-intl. I expose a folder say 'i18n' along with the distribution folder for my library. For the locales that need to be supported, I have placed a json file which has the translations as key/value pairs, inside this folder (for example to support french, I have i18n/fr.json)

No if the user who use this library want to add support for additional locale, he can place an additional json file in this folder with translated text.( for example to support german, user can create i18n/de.json)

But since in the library, I have not included imports for de locale data or have included deLocaleData in addLocaleData(), I ll encounter missing locale data for 'de' error. 

Here where, it will be very useful for me if there is a provision to add the locale data dynamically. I can even ship the localedata folder along with the library and it will be nice to have a load on demand feature. 
(say load delocaleData only on DE locale)

**Comment by ericf** [comment]
By dynamic do you mean to allow locale-switching in the browser and that you want to fetch `react-intl/locale-data/de.js` over the network async?

I would recommend doing something similar to what React Intl does for loading locale data. It's not React Intl that actually needs the locale data, it's the underlying `intl-messageformat` and `intl-relativeformat` libs. `addLocaleData()`, simply calls `addLocaleData()` on this libs and passes the data through. You can create _another_ wrapper around this for your lib which you create `locale-data/de.js`, etc. modules which include locale data for React Intl and your lib and add them to a registry.

From the end user's perspective it could look very similar to how React Intl's locale data is added to its registry:

```js
import {addLocaleData} from 'react-intl';
import de from 'react-intl/locale-data/de';
addLocaleData(de);
```
or:
```js
<script src="https://unpkg.com/react-intl@latest/dist/react-intl.min.js"></script>
<script src="https://unpkg.com/react-intl@latest/locale-data/de.js"></script>
<script>
    ReactIntl.addLocaleData(ReactIntlLocaleData.de);
</script>
```
https://github.com/yahoo/react-intl/wiki#loading-locale-data

Checkout the `locale-data-registry.js` file for how this is done: https://github.com/yahoo/react-intl/blob/master/src/locale-data-registry.js

**Comment by SanthoshHari912** [comment]
Thanks for the suggestions. I followed as mentioned in locale-data-registry.js.
So say if my locale is 'de', I did the following

			let localeData = {
				locale: navigator.language;
			};
			IntlMessageFormat.__addLocaleData(localeData);
			IntlRelativeFormat.__addLocaleData(localeData);
but I get the below error,
Error: Locale data added to IntlMessageFormat is missing a `pluralRuleFunction` for :de

**Comment by peramsathyam** [comment]
@SanthoshHari912 : if you are not using `browserify` for your building and using `webpack`

you can do : 
```
    import { addLocaleData } from 'react-intl';
    const possibleLocale = navigator.language || 'en';
    addLocaleData(require(`react-intl/locale-data/${possibleLocale}`));
```

You can also try dynamic loading with `System.import` promise similar to `require.ensure`

```
System.import(`react-intl/locale-data/${possibleLocale}`).then((localeData) => {
  addLocaleData(localeData);
});
```

**Comment by SanthoshHari912** [comment]
@peramsathyam Thank you .This is exactly what I was trying to achieve. It worked like charm. Just a small change I have to do the code above.

    import { addLocaleData } from 'react-intl';
    const possibleLocale = navigator.language.split('-')[0] || 'en';
    addLocaleData(require(`react-intl/locale-data/${possibleLocale}`));

**Comment by arvinsim** [comment]
```
import { addLocaleData } from 'react-intl';
const possibleLocale = navigator.language.split('-')[0] || 'en';
addLocaleData(require(`react-intl/locale-data/${possibleLocale}`));
```

This I think does it dynamically from navigator language, but seems to grab only the first bit before hyphen so would not work for languages like “es-BO” or “ru-MI” so we just tweak that bit for the second half

**Comment by StorytellerCZ** [comment]
@arvinsim Yes, the locales are based on languages, not on countries. If you have to have country based locale you need to implement it yourself.

**Comment by sergioviniciuss** [comment]
I succeeded to dynamically import my localedata with this code:
``` javascript
const locale = Locale.getLocale();
addLocaleData(require(`react-intl/locale-data/${locale}`));
```
But I don't want to use require, in order to lazy load some data, So I tried this:
const locale = Locale.getLocale();
``` javascript
(async () => {
  const localeData = await import(`react-intl/locale-data/${locale}`);
  addLocaleData(localeData);
})();
```
and it didn't work. Locale is not being loaded. Any ideas what can I possibly be doing wrong?

**Comment by kylehalleman** [comment]
@sergioviniciuss I have encountered the same issue. In `componentDidMount` I check the browser language and if it's not English (which the translation file is bundled into the app), then I will dynamically load both the localeData and the translation file using `import()`. The correct data is returned and I use `addLocaleData(data.default)` and then I get missing message errors. (I tried `addLocaleData(data)` and it would complain the localeData wasn't added for that language.
